### DISCUSSION TITLE
How to define a half normal detection function with Cosine adjustment

### DISCUSSION THREAD

**Discussion by lvclarke** [inlabru-org/inlabru#195] [General]
Hi, 

I have been looking at the dolphin and gorilla examples which always specify a half normal detection function within the spatial model:
hn <- function(distance, lsig) {
  [exp](https://rdrr.io/r/base/Log.html)(-0.5 * (distance / [exp](https://rdrr.io/r/base/Log.html)(lsig))^2)
}

However, the detection probability that best fits my data is half normal with cosine adjustment. Is there a way that this can be integrated into my code?

Many thanks,
Lucy

**Comment by finnlindgren** [comment]
In the distance sampling article, there's also an example of using a hazard-rate detection model (https://inlabru-org.github.io/inlabru/articles/web/2d_lgcp_distancesampling.html#hazard-rate-detection-function), by defining a different detection probability function. Unless there's some specific mathematical problem with your model, then you should be able to implement it in the same way as the two examples in the article.

**Reply by lvclarke** [reply]
Hi Finn, Thanks for such a a quick reply!

I did see the hazard-rate detection model - I'm just confused about the correct formula to apply the cosine adjustment? Apols, might be a silly question that reflects my lack of stats background!

**Reply by finnlindgren** [reply]
Do you have a mathematical expression for this "cosine adjustment"? (Or if not, then some other description that defines what the model is.) I'm not familiar with this model, so I'm not sure what you need to change in the code.

**Reply by finnlindgren** [reply]
E.g. the half-normal model is
$g(d;\sigma) = \exp(-(d/\sigma)^2/2)$

**Reply by finnlindgren** [reply]
And the hazard-rate model in the vignette is
$g(s;\sigma)=1-\exp(-(d/\sigma)^{-1})$
and more generally
$g(s;\sigma,\gamma)=1-\exp(-(d/\sigma)^{-\gamma})$

**Reply by lvclarke** [reply]
<img width="349" alt="image" src="https://github.com/inlabru-org/inlabru/assets/135431699/b7322fe4-be6e-4454-a519-b5d9f6910de5">

Thanks very much - I used the distance package with this code:
hn.cos <- ds(data=data2, key="hn", adjustment="cos")
Which is listed as "Half-normal with cosine adjustment terms of order 2,3,4" and gives this plot. Unfortunately don't have much else to go off and wondering if it is feasible

**Reply by finnlindgren** [reply]
I see. Perhaps the model is described in one of the books,
```
Buckland, S.T., Anderson, D.R., Burnham, K.P., Laake, J.L., Borchers, D.L., and Thomas, L. (2001). Distance Sampling. Oxford University Press. Oxford, UK.

Buckland, S.T., Anderson, D.R., Burnham, K.P., Laake, J.L., Borchers, D.L., and Thomas, L. (2004). Advanced Distance Sampling. Oxford University Press. Oxford, UK.
```
My guess is that they're either multiplying or adding the cosine functions, but might need to dig into their source code to find out, as the documentation doesn't say...

**Reply by finnlindgren** [reply]
Found a general explanation of their "adjusted" models here: https://workshops.distancesampling.org/online-course/lecturepdfs/Ch1/L1-4%20Choosing%20a%20Detection%20Function.pdf
It should be simple to implement (but with too many parameters and flexibility there may be practical issues when doing the estimation). I'll reply more later.

**Reply by lvclarke** [reply]
Thanks Finn - will take a look!

**Comment by finnlindgren** [comment]
The model the pdf claims they use is
$g(d;\sigma) (1 + \text{series}(d))$ (normalised to 1 at $d=0$)
but that's highly problematic as it doesn't guarantee a positive probability. It also doesn't guarantee that the probability stays below $1$, but that's less of a problem, as it can be used as an indication that the probability at distance zero should be less than 1; i.e. that a different normalisation is needed.

As an alternative, that for small series adjustments will be very similar to their model, I would suggest
$g(d;\sigma) \exp(\text{series}(d))$
(also normalised at distance zero). The first-order Taylor expansion of the exponential is the same as their model.
It also has the advantage of having the series coefficients entering linearly in the log-probability;
$\log(\text{probability}) = \log(g(d;\sigma)) + \text{series}(d) + \text{normalisation}$.
This is still linear in the series coefficients after normalisation.

Something like this:

Model definition
```
# cosine series, orders 1, 2, 3, 4, up to length(series_coef)
log_hn_cos_adj <- function(distance, lsig, series_coef, W) {
  log_prob <- -0.5 * (distance / exp(lsig))^2)
  log_prob0 <- 0
  series <- 0
  for (k in seq_along(series_coef)) {
    series <- series + cos(distance / W * pi * k) * series_coef[k]
  }
  series0 <- sum(series_coef)
  log_prob - log_prob0 + series - series0
}
```
Definitions for a model with orders 2,3, and 4 adjustments:
```
components <- ... +
  lsig(1) +
  series_coef(1:3, model = "iid", hyper = list(prec = list(initial=log(10), fixed = TRUE)))
formula <- ... +
  log_hn_cos_adj(distance, lsig = lsig, series_coef = c(0, series_coef), W = half_width)
```
